Feat: Kakao OAuth 추가

saebyeok0306 opened the pull request:
## Kakao OAuth 추가

1. @dasd412 님이 작업함.
3. 로그인 클릭시 Kakao 로그인페이지로 이동하여 로그인을 진행함.
4. checkLogin함수로 매번 로그인 상태를 체크하고 AccessToken이 만료된 경우 RefreshToken으로 갱신함.
5. Axios Client를 Token이 필요한 상황에 맞춰서 추가함.

## Axios Client 오류 수정

1. Create Axios로 Client를 생성할 때, 쿠키로 Token도 함께 보내도록 header 설정을 포함하고 있음.
2. 간과하고 있던 부분은 Create Axios는 `생성된 Instance` 였음.
3. 처음 생성될 때 당시에 쿠키값을 설정하고 갱신이 되지 않아서 RefreshToken이 제대로 작동하지 않음.
4. 해당 부분의 코드는 함수를 통해 header를 업데이트하는 식으로 수정했고, 네이밍을 Client에서 Instance로 변경함.

## OAuth를 통해 받아온 사용자 정보 Recoil에서 관리

1. userAtom에서 로그인여부, 닉네임, 프로필사진(url)을 관리함.
2. 로그아웃하게 될 경우, 해당 상태변수들도 초기화하도록 관리하고 있음.

## JWT 컴포넌트

1. OAuth 상태관리를 위한 JWT 컴포넌트를 만듬.
2. JWT 컴포넌트는 App.js에서 실행되며, 이는 페이지가 새로 렌더링될 때마다 딱 1번씩 갱신됨.
3. JWT 컴포넌트에서 비로그인 상태일 때, 접근할 수 없는 Path에 대한 통제도 담당하고 있음.

## 사용자 정보를 표기하기 위한 모바일용 햄버거 버튼 추가

1. Header에 모든 버튼과 사용자 정보를 표기할 경우 모바일에서 UI가 깨지게 됨.
2. 모바일 사용자를 위해 모바일에서는 햄버거 버튼 방식으로 메뉴를 소환하는 식으로 개선됨.

## Main 페이지 수정

1. 로그인했을 때에만 면접시작 버튼이 나오도록 변경함.
2. 로그인하지 않은 상태일 때는 카카오 로그인 버튼이 나타남.

## 사용자 정보 입력단계에서 사용자 이름 항목 제거

1. 사용자 이름은 OAuth를 통해 얻을 수 있으므로, 입력단계에서 사용자 이름 입력칸을 삭제함.
2. API를 호출할 땐 Recoil에서 관리하는 사용자 이름을 불러와서 사용함.

## Footer 개선

1. 허전해보여서 조금 더 채워봤습니다.

dasd412 committed:
feat : oauth page component 추가

dasd412 committed:
style : css 파일 및 폰트, 버튼 이미지 추가

dasd412 committed:
style : react 문법에 맞게 스타일 적용 + 로딩 바는 애니메이션 css에 옮김.

dasd412 committed:
feat : 1차적으로 로그인 로직 추가.

dasd412 committed:
merge into feat-oauth

dasd412 committed:
refacor : js -> react로 리팩토링

dasd412 committed:
feat : login callback 페이지 완성

dasd412 committed:
feat : callback page 용 라우팅 추가

dasd412 committed:
Merge branch 'dev' into feat-oauth

dasd412 committed:
fix : login url 변경 및 썸네일 변수 수정

dasd412 committed:
feat : logout 로직 완성

dasd412 committed:
refactor : 완성한 메서드에 남긴  console log 삭제

dasd412 committed:
feat : 로그아웃 이후 url 이동 추가 및 리프레시 토큰 로직 추가

dasd412 committed:
feat : refresh token 로직까지 타는 것 완료

dasd412 committed:
fix : refresh token 로직 invalidate

dasd412 committed:
feat : api 요청에 액세스 토큰 포함 성공!!!

dasd412 committed:
feat : refresh token 로직 완료~!

dasd412 committed:
fix & refactor : 인증용 axios 따로 공통 js로 뺌

dasd412 committed:
fix : oauth 로직 헤더로 이관

dasd412 committed:
feat & rename : api_client_token으로 이름 변경 및 입력 페이지 이동 시 인가 추가

dasd412 committed:
feat : 입력 페이지에도 리프레시 체크 로직이 있음. 그런데 헤더랑 중복되는데, 이를 어떻게 해야 할 지 모르겠다.

dasd412 committed:
fix : 회원 가입 안할 시 에러 나던 것 예외 처리 추가

saebyeok0306 committed:
refactor: API Instance로 이름 변경 및 jwt 토큰헤더 설정

Signed-off-by: JangSeHun <westreed@naver.com>

saebyeok0306 committed:
refactor: JWT를 최상위 컴포넌트로 이동 및 UserAtom 추가

Signed-off-by: JangSeHun <westreed@naver.com>

saebyeok0306 committed:
feat: 로그인, 로그아웃 버튼 추가를 위해 모바일 햄버거버튼 추가

Signed-off-by: JangSeHun <westreed@naver.com>

saebyeok0306 committed:
refactor: 로그인했을 때에만 면접시작 버튼이 나타도록 변경

Signed-off-by: JangSeHun <westreed@naver.com>

saebyeok0306 committed:
refactor: footer 디자인 개선

Signed-off-by: JangSeHun <westreed@naver.com>

saebyeok0306 committed:
refactor: 코드 정리

Signed-off-by: JangSeHun <westreed@naver.com>

saebyeok0306 committed:
refactor: OAuth를 통해 받은 사용자이름으로 Input을 대신하도록 변경

Signed-off-by: JangSeHun <westreed@naver.com>

ssfic3380 commented on src/JWT.js:
```diff
@@ -0,0 +1,75 @@
+  };
+
+  const refreshTokenInChecking = async () => {
+    return jwt_refresh_api()
```
이렇게 바로 api를 return하는 건 api 호출을 함수화 하기 위함인가요?

saebyeok0306 replied:
Promise 문법인데 저렇게 return을 하면 then나 catch에서 return한 값을 받아와서 그대로 refreshTokenInChecking의 결과로 리턴하는 식이에요. 그런데 사실 딱히 return하고 있는 것도 없고 return값을 쓰는 것도 없어서 큰 의미는 없다고 보시면 될 것 같습니다.

다른 코드 복붙하면서 그대로 쓴 것 같아요 ㅋㅋ

ssfic3380 commented on src/api/api_instance.js:
```diff
@@ -0,0 +1,38 @@
+  return cookieValue;
+};
+
+export const API_INSTANCE = axios.create({
```
기존에 단순히 apiClient만 이용하던 것을 이렇게 각각 헤더의 값을 다르게 지정해 놓은 것은 api마다 인증(jwt)이 필요하기도 하고, 필요 없는 api도 있기도 해서 라고 이해했습니다. 맞게 이해한걸까요?

만약 그렇다면 api 코드에 apiClient.defaults.headers[...] 이렇게 각각 지정해주는 것과 차이가 뭔가요? 재사용을 위함인가요?

saebyeok0306 replied:
axios.create는 말그대로 객체생성이라서 이때 header값을 넣어두면 생성될 때 당시의 값만 그대로 들어갑니다. 하지만 저희가 헤더에 넣는 Token 값은 일정주기마다 값이 계속 바뀌기 때문에 갱신이 되지 않으면 문제가 생기게 됩니다.

그래서 axios 객체는 하나만 생성해두고, 따로 함수를 통해서 axios 객체의 header를 수정해서 갱신한다고 보시면 됩니다.

ssfic3380 commented on src/components/footer.jsx:
```diff
@@ -6,8 +6,13 @@ function Footer() {
+            <div className={`${style.footer_list}`}>
+              <h5>모두의 인터뷰</h5>
+              <span>모뷰 | Team : Architect</span>
+              <span>서울특별시 강남구 테헤란로 311 아남타워 7층</span>
```
ㄷㄷ 민원 다 소마 센터로 찾아올 예정

ssfic3380 commented on src/components/header.jsx:
```diff
@@ -20,50 +20,131 @@ function InterviewMenu() {
 
   const handleButtonClick = (e) => {
     e.preventDefault();
+    setIsClickedHamburger(false);
```
'클릭 된 햄버거인가'도 맞긴 한데 setIsHamburgerClicked 가 좀 더 자연스러울 것 같습니다.

saebyeok0306 replied:
햄버거로 맞고 싶으십니까? ㅋㅋ 언젠가 반영할게요...

ssfic3380 reviewed (approved):
수고 많으셨습니다.
만약 사용자 인증이 필요한 api라면 api_instance를 import 해서 apiInstanceForAccess를 활용하면 될까요?
그리고 apiInstanceForRefresh는 저희가 따로 이용할 일이 있을지 궁금합니다.

saebyeok0306 commented:
> 수고 많으셨습니다. 만약 사용자 인증이 필요한 api라면 api_instance를 import 해서 apiInstanceForAccess를 활용하면 될까요? 그리고 apiInstanceForRefresh는 저희가 따로 이용할 일이 있을지 궁금합니다.

apiInstanceForRefresh는 실제로 JWT 컴포넌트에서 Access Token이 만료되었을 때 갱신하는 용도로 쓰고 있어요.

saebyeok0306 committed:
refactor: Axios Intercepter 사용

토큰을 필요로 하는 api 요청은 중간에 가로채어 header에 토큰을 넣어주고,
해당 토큰이 만료된 경우 refresh token과 함께 access token을 갱신 후 다시 요청을 보냄.

Signed-off-by: JangSeHun <westreed@naver.com>

saebyeok0306 committed:
remove: 기존의 JWT 컴포넌트 삭제

axios intercepter로 기존 로직을 대체함.

Signed-off-by: JangSeHun <westreed@naver.com>

saebyeok0306 committed:
refactor: 로그인할 때 사용자정보를 바로 갱신하도록 변경

Signed-off-by: JangSeHun <westreed@naver.com>

saebyeok0306 committed:
chore: mem 라이브러리 추가

함수의 return값을 일정시간 동안 cache로 저장하여
return 하는 라이브러리. refresh api 요청이 동시에 왔을 때, 같은 token을 돌려주기 위해 사용함.

Signed-off-by: JangSeHun <westreed@naver.com>

saebyeok0306 committed:
refactor: Axios Interceptor에서 React Hook을 쓸 수 있게 변경

Signed-off-by: JangSeHun <westreed@naver.com>

saebyeok0306 committed:
feat: react-cookie 라이브러리 추가

Signed-off-by: JangSeHun <westreed@naver.com>

saebyeok0306 committed:
refactor: Google Analytics Pageview 개선

Signed-off-by: JangSeHun <westreed@naver.com>
